secrecy/src/lib.rs

```diff
@@ -15,239 +13,183 @@
 /// (e.g. passwords, cryptographic keys, access tokens or other credentials)
 ///
-/// Access to the secret inner value occurs through the [`ExposeSecret`] trait,
-/// which provides an `expose_secret()` method for accessing the inner secret
-/// value.
-pub struct Secret<S>
-where
-    S: Zeroize,
-{
-    /// Inner secret value
-    inner_secret: S,
+/// Access to the secret inner value occurs through the [`ExposeSecret`]
+/// or [`ExposeSecretMut`] traits, which provide methods for accessing the inner secret value.
+pub struct SecretBox<S: Zeroize> {
+    inner_secret: Box<S>,
 }
 
-impl<S> Secret<S>
-where
-    S: Zeroize,
-{
-    /// Take ownership of a secret value
-    pub fn new(secret: S) -> Self {
-        Secret {
-            inner_secret: secret,
+impl<S: Zeroize> Zeroize for SecretBox<S> {
+    fn zeroize(&mut self) {
+        self.inner_secret.as_mut().zeroize()
+    }
+}
+
+impl<S: Zeroize> Drop for SecretBox<S> {
+    fn drop(&mut self) {
+        self.zeroize()
+    }
+}
+
+impl<S: Zeroize> ZeroizeOnDrop for SecretBox<S> {}
+
+impl<S: Zeroize> From<Box<S>> for SecretBox<S> {
+    fn from(source: Box<S>) -> Self {
+        Self::new(source)
+    }
+}
+
+impl<S: Zeroize> SecretBox<S> {
+    /// Create a secret value using a pre-boxed value.
+    pub fn new(boxed_secret: Box<S>) -> Self {
+        Self {
+            inner_secret: boxed_secret,
         }
     }
 }
 
-impl<S> ExposeSecret<S> for Secret<S>
-where
-    S: Zeroize,
-{
-    fn expose_secret(&self) -> &S {
-        &self.inner_secret
+impl<S: Zeroize + Default> SecretBox<S> {
+    /// Create a secret value using a function that can initialize the vale in-place.
+    pub fn new_with_mut(ctr: impl FnOnce(&mut S)) -> Self {
+        let mut secret = Self::default();
+        ctr(secret.expose_secret_mut());
+        secret
     }
 }
 
-impl<S> From<S> for Secret<S>
-where
-    S: Zeroize,
-{
-    fn from(secret: S) -> Self {
-        Self::new(secret)
+impl<S: Zeroize + Clone> SecretBox<S> {
+    /// Create a secret value using the provided function as a constructor.
+    ///
+    /// The implementation makes an effort to zeroize the locally constructed value
+    /// before it is copied to the heap, and constructing it inside the closure minimizes
+    /// the possibility of it being accidentally copied by other code.
+    ///
+    /// **Note:** using [`Self::new`] or [`Self::new_with_mut`] is preferable when possible,
+    /// since this method's safety relies on empyric evidence and may be violated on some targets.
+    pub fn new_with_ctr(ctr: impl FnOnce() -> S) -> Self {
+        let mut data = ctr();
+        let secret = Self {
+            inner_secret: Box::new(data.clone()),
+        };
+        data.zeroize();
+        secret
+    }
+
+    /// Same as [`Self::new_with_ctr`], but the constructor can be fallible.
+    pub fn new_with_fallible_ctr<E>(ctr: impl FnOnce() -> Result<S, E>) -> Result<Self, E> {
```

Comment:
I think this is a bit more idiomatic:

```suggestion
    pub fn try_new_with_ctr<E>(ctr: impl FnOnce() -> Result<S, E>) -> Result<Self, E> {
```

Reply:
Fixed.